File: src/com/checkers/board/Cell.java
package com.checkers.board;

import com.checkers.exceptions.CanNotMoveException;
import com.checkers.utils.Colour;

public class Cell {
    private int col;
    private int row;
    private Colour colour;
    private Piece piece;
    private CheckerBoard board;

    public Cell(int _row, int _col, CheckerBoard _board) {
        col = _col;
        row = _row;
        board = _board;
        piece = null;
        if ((col + row) % 2 == 0) colour = Colour.BLACK;
        else colour = Colour.WHITE;
    }

    public int getCol() {
        return col;
    }

    public int getRow() {
        return row;
    }

    public Piece getPiece() throws NullPointerException {
        return piece;
    }

    public void setPiece(Piece piece) {
        this.piece = piece;
    }

    public Colour getColour() {
        return colour;
    }

    public String getMap() {
        String result = "";
        result += (char) ('a' + getCol() - 1);
        result += getRow();
        return result;
    }

    @Override
    public String toString() {
        try {
            return piece.toString();
        } catch (NullPointerException ex) {
            if (colour == Colour.BLACK) {
                return "*";
            } else {
                return "^";
            }
        }
    }

    public int diff(Cell second) throws CanNotMoveException {
        if (Math.abs(getCol() - second.getCol()) == Math.abs(getRow() - second.getRow())) {
            return Math.abs(getCol() - second.getCol());
        }
        throw new CanNotMoveException();
    }

    public Cell getNear(int diffRow, int diffCol) {
        return board.getCell(getRow() + diffRow, getCol() + diffCol);
    }

    public Cell between(Cell another, CheckerBoard board) {
        int col = (getCol() + another.getCol()) / 2;
        int row = (getRow() + another.getRow()) / 2;
        return board.getCell(row - 1, col - 1);
    }
}

File: src/main/kotlin/com/checkers/board/Cell.kt
package com.checkers.board

import com.checkers.utils.Color
import com.checkers.utils.isCoordinatesExists
import kotlin.math.abs

data class Cell(
    val row: Int,
    val column: Int,
    val board: Board,
) {

    var piece: Piece? = null
        set(value) {
            field = value
            if (value != null) {
                value.cell = this
            }
        }

    val color: Color
        get() = if ((column + row) % 2 == 0) {
            Color.BLACK
        } else {
            Color.WHITE
        }

    fun getMap() =
        toString()

    fun getNear(diffRow: Int, diffColumn: Int): Cell {
        val rowToFind = row + diffRow
        val columnToFind = column + diffColumn
        if (isCoordinatesExists(rowToFind, columnToFind)) {
            return board.getCell(rowToFind, columnToFind)
        }

        throw ArrayIndexOutOfBoundsException()
    }

    fun diff(second: Cell): Int {
        if (abs(column - second.column) == abs(row - second.row)) {
            return abs(column - second.column)
        }

        return -1
    }

    fun between(another: Cell, board: Board): Cell {
        val colToFind = (column + another.column) / 2
        val rowToFind = (row + another.row) / 2
        return board.getCell(rowToFind, colToFind)
    }

    fun isEmpty(): Boolean =
        piece == null

    override fun toString(): String {
        val letter = ('a' + column - 1)
        return letter.toString() + row
    }
}